File: src/main/java/io/vertx/ext/apex/addons/StaticServer.java
/*
 * Copyright 2014 Red Hat, Inc.
 *
 *  All rights reserved. This program and the accompanying materials
 *  are made available under the terms of the Eclipse Public License v1.0
 *  and Apache License v2.0 which accompanies this distribution.
 *
 *  The Eclipse Public License is available at
 *  http://www.eclipse.org/legal/epl-v10.html
 *
 *  The Apache License v2.0 is available at
 *  http://www.opensource.org/licenses/apache2.0.php
 *
 *  You may elect to redistribute this code under either of these licenses.
 */

package io.vertx.ext.apex.addons;

import io.vertx.codegen.annotations.VertxGen;
import io.vertx.core.Handler;
import io.vertx.ext.apex.core.RoutingContext;
import io.vertx.ext.apex.addons.impl.StaticServerImpl;

/**
 * @author <a href="http://tfox.org">Tim Fox</a>
 */
@VertxGen
public interface StaticServer extends Handler<RoutingContext> {

  static final String DEFAULT_WEB_ROOT = "webroot";
  static final boolean DEFAULT_FILES_READ_ONLY = true;
  static final long DEFAULT_MAX_AGE_SECONDS = 86400; // One day
  static final boolean DEFAULT_CACHING_ENABLED = true;
  static final boolean DEFAULT_DIRECTORY_LISTING = false;
  static final String DEFAULT_DIRECTORY_TEMPLATE = "apex-directory.html";
  static final boolean DEFAULT_INCLUDE_HIDDEN = true;
  static final long DEFAULT_CACHE_ENTRY_TIMEOUT = 30000; // 30 seconds
  static final String DEFAULT_INDEX_PAGE = "/index.html";
  static final int DEFAULT_MAX_CACHE_SIZE = 10000;
  static final boolean DEFAULT_ALWAYS_ASYNC_FS = false;
  static final boolean DEFAULT_ENABLE_FS_TUNING = true;
  static final long DEFAULT_MAX_AVG_SERVE_TIME_NS = 1000000; // 1ms

  static StaticServer staticServer(String root) {
    return new StaticServerImpl(root);
  }

  static StaticServer staticServer() {
    return new StaticServerImpl();
  }

  @Override
  void handle(RoutingContext context);

  StaticServer setWebRoot(String webRoot);

  StaticServer setFilesReadOnly(boolean readOnly);

  StaticServer setMaxAgeSeconds(long maxAgeSeconds);

  StaticServer setCachingEnabled(boolean enabled);

  StaticServer setDirectoryListing(boolean directoryListing);

  StaticServer setIncludeHidden(boolean includeHidden);

  StaticServer setCacheEntryTimeout(long timeout);

  StaticServer setIndexPage(String indexPage);

  StaticServer setMaxCacheSize(int maxCacheSize);

  StaticServer setAlwaysAsyncFS(boolean alwaysAsyncFS);

  StaticServer setEnableFSTuning(boolean enableFSTuning);

  StaticServer setMaxAvgServeTimeNs(long maxAvgServeTimeNanoSeconds);

  StaticServer setDirectoryTemplate(String directoryTemplate);

}

File: src/main/resources/vertx-apex-addons-js/static_server.js
/*
 * Copyright 2014 Red Hat, Inc.
 *
 * Red Hat licenses this file to you under the Apache License, version 2.0
 * (the "License"); you may not use this file except in compliance with the
 * License.  You may obtain a copy of the License at:
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */

/** @module vertx-apex-addons-js/static_server */
var utils = require('vertx-js/util/utils');
var RoutingContext = require('vertx-apex-core-js/routing_context');

var io = Packages.io;
var JsonObject = io.vertx.core.json.JsonObject;
var JStaticServer = io.vertx.ext.apex.addons.StaticServer;

/**

 @class
*/
var StaticServer = function(j_val) {

  var j_staticServer = j_val;
  var that = this;

  /**

   @public
   @param context {RoutingContext} 
   */
  this.handle = function(context) {
    var __args = arguments;
    if (__args.length === 1 && typeof __args[0] === 'object' && __args[0]._jdel) {
      j_staticServer.handle(context._jdel);
    } else utils.invalidArgs();
  };

  /**

   @public
   @param webRoot {string} 
   @return {StaticServer}
   */
  this.setWebRoot = function(webRoot) {
    var __args = arguments;
    if (__args.length === 1 && typeof __args[0] === 'string') {
      return new StaticServer(j_staticServer.setWebRoot(webRoot));
    } else utils.invalidArgs();
  };

  /**

   @public
   @param readOnly {boolean} 
   @return {StaticServer}
   */
  this.setFilesReadOnly = function(readOnly) {
    var __args = arguments;
    if (__args.length === 1 && typeof __args[0] ==='boolean') {
      return new StaticServer(j_staticServer.setFilesReadOnly(readOnly));
    } else utils.invalidArgs();
  };

  /**

   @public
   @param maxAgeSeconds {number} 
   @return {StaticServer}
   */
  this.setMaxAgeSeconds = function(maxAgeSeconds) {
    var __args = arguments;
    if (__args.length === 1 && typeof __args[0] ==='number') {
      return new StaticServer(j_staticServer.setMaxAgeSeconds(maxAgeSeconds));
    } else utils.invalidArgs();
  };

  /**

   @public
   @param enabled {boolean} 
   @return {StaticServer}
   */
  this.setCachingEnabled = function(enabled) {
    var __args = arguments;
    if (__args.length === 1 && typeof __args[0] ==='boolean') {
      return new StaticServer(j_staticServer.setCachingEnabled(enabled));
    } else utils.invalidArgs();
  };

  /**

   @public
   @param directoryListing {boolean} 
   @return {StaticServer}
   */
  this.setDirectoryListing = function(directoryListing) {
    var __args = arguments;
    if (__args.length === 1 && typeof __args[0] ==='boolean') {
      return new StaticServer(j_staticServer.setDirectoryListing(directoryListing));
    } else utils.invalidArgs();
  };

  /**

   @public
   @param includeHidden {boolean} 
   @return {StaticServer}
   */
  this.setIncludeHidden = function(includeHidden) {
    var __args = arguments;
    if (__args.length === 1 && typeof __args[0] ==='boolean') {
      return new StaticServer(j_staticServer.setIncludeHidden(includeHidden));
    } else utils.invalidArgs();
  };

  /**

   @public
   @param timeout {number} 
   @return {StaticServer}
   */
  this.setCacheEntryTimeout = function(timeout) {
    var __args = arguments;
    if (__args.length === 1 && typeof __args[0] ==='number') {
      return new StaticServer(j_staticServer.setCacheEntryTimeout(timeout));
    } else utils.invalidArgs();
  };

  /**

   @public
   @param indexPage {string} 
   @return {StaticServer}
   */
  this.setIndexPage = function(indexPage) {
    var __args = arguments;
    if (__args.length === 1 && typeof __args[0] === 'string') {
      return new StaticServer(j_staticServer.setIndexPage(indexPage));
    } else utils.invalidArgs();
  };

  /**

   @public
   @param maxCacheSize {number} 
   @return {StaticServer}
   */
  this.setMaxCacheSize = function(maxCacheSize) {
    var __args = arguments;
    if (__args.length === 1 && typeof __args[0] ==='number') {
      return new StaticServer(j_staticServer.setMaxCacheSize(maxCacheSize));
    } else utils.invalidArgs();
  };

  /**

   @public
   @param alwaysAsyncFS {boolean} 
   @return {StaticServer}
   */
  this.setAlwaysAsyncFS = function(alwaysAsyncFS) {
    var __args = arguments;
    if (__args.length === 1 && typeof __args[0] ==='boolean') {
      return new StaticServer(j_staticServer.setAlwaysAsyncFS(alwaysAsyncFS));
    } else utils.invalidArgs();
  };

  /**

   @public
   @param enableFSTuning {boolean} 
   @return {StaticServer}
   */
  this.setEnableFSTuning = function(enableFSTuning) {
    var __args = arguments;
    if (__args.length === 1 && typeof __args[0] ==='boolean') {
      return new StaticServer(j_staticServer.setEnableFSTuning(enableFSTuning));
    } else utils.invalidArgs();
  };

  /**

   @public
   @param maxAvgServeTimeNanoSeconds {number} 
   @return {StaticServer}
   */
  this.setMaxAvgServeTimeNs = function(maxAvgServeTimeNanoSeconds) {
    var __args = arguments;
    if (__args.length === 1 && typeof __args[0] ==='number') {
      return new StaticServer(j_staticServer.setMaxAvgServeTimeNs(maxAvgServeTimeNanoSeconds));
    } else utils.invalidArgs();
  };

  // A reference to the underlying Java delegate
  // NOTE! This is an internal API and must not be used in user code.
  // If you rely on this property your code is likely to break if we change it / remove it without warning.
  this._jdel = j_staticServer;
};

/**

 @memberof module:vertx-apex-addons-js/static_server

 @return {StaticServer}
 */
StaticServer.staticServer = function() {
  var __args = arguments;
  if (__args.length === 0) {
    return new StaticServer(JStaticServer.staticServer());
  }else if (__args.length === 1 && typeof __args[0] === 'string') {
    return new StaticServer(JStaticServer.staticServer(__args[0]));
  } else utils.invalidArgs();
};

// We export the Constructor function
module.exports = StaticServer;